### DISCUSSION TITLE
关于CreateLoad的用法

### DISCUSSION THREAD

**Discussion by Aurora-1412** [arcsysu/SYsU-lang#48] [Q&A]
当我们获取一个变量，例如a[1][2]时，首先在符号表里查询a，查到之后返回一个指向最外层的指针。

然后我们需要逐层向内CreateInBoundsGEP，直到读取完a[1][2]，是不是中间都不需要CreateLoad啊
另外CreateLoad是否仅仅用于左值转右值的时候呢

**Comment by Yorkking** [comment]
两种情况：
- 左值转右值
- 本身是右值

**Reply by Aurora-1412** [reply]
> 两种情况：
> 
> * 左值转右值
> * 本身是右值

谢谢学长，顺便再问一下，出现报错
'''
static llvm::GetElementPtrInst *llvm::GetElementPtrInst::Create(llvm::Type *, llvm::Value *, ArrayRef<llvm::Value *>, const llvm::Twine &, llvm::Instruction *): Assertion `PointeeType == cast<PointerType>(Ptr->getType()->getScalarType())->getElementType()' failed.
'''
可能是什么原因呢，是在‘ Value* arr_ptr = Builder.CreateInBoundsGEP(arr_type, RetVal, idxList);’这一句出错的，其中RetVal的值为：
%7 = getelementptr inbounds [4 x [2 x i32]], [4 x [2 x i32]]* %a, i32 0, i32 0
arr_type的值为：[4 x [2 x i32]]
定义idxList的语句：'Value* idxList[2] = {ConstantInt::get(TheContext,APInt(32,"0",10)),int_val};' ,其中int_val 为 i32 0

**Comment by Yorkking** [comment]
好像你这个 arr_type 有问题，应该是 [2 x i32] 吧。另外， Builder.CreateInBoundsGEP 可以不用传入第一个参数 type 类型。

**Comment by Aurora-1412** [comment]
谢谢学长，去掉arr_type就不报错了



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2022年5月22日(周日) 晚上8:11
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [arcsysu/SYsU-lang] 关于CreateLoad的用法 (Discussion #48)




 
好像你这个 arr_type 有问题，应该是 [2 x i32] 吧。另外， Builder.CreateInBoundsGEP 可以不用传入第一个参数 type 类型。
